### ISSUE TITLE
Skia rendering of labels is in world coordinates for some rendereres

### ISSUE THREAD

**Issue by Kortdag** [pauldendulk/Mapsui#275]
Bug #262 

I had a similar issue on iOS which also uses the Skia renderer. After looking into to source code there is a critical bug in the Polygon and LineString renderer (and maybe others) compared to the PointRenderer. 

The point renderer uses WorldToScreen transformation befor drawing on the canvas. Like so.
```static class PointRenderer
    {
        // todo: 
        // try to remove the feature argument. LabelStyle should already contain the feature specific text
        // The visible feature iterator should create this LabelStyle
        public static void Draw(SKCanvas canvas, IViewport viewport, IStyle style, IFeature feature, 
            IGeometry geometry, SymbolCache symbolCache)
        {
            var point = geometry as Point;
            var destination = viewport.WorldToScreen(point);

            if (style is LabelStyle)              // case 1) LabelStyle
            {
                LabelRenderer.Draw(canvas, (LabelStyle) style, feature, (float) destination.X, (float) destination.Y);
            }
```

The polygon and LineString renderer does no transform from world to screen so the canvas anchor is in world coordinates, like so
```
    public static class LineStringRenderer
    {
        public static void Draw(SKCanvas canvas, IViewport viewport, IStyle style, IFeature feature, IGeometry geometry)
        {
            if (style is LabelStyle)
            {
                var center = geometry.GetBoundingBox().GetCentroid();
                LabelRenderer.Draw(canvas, (LabelStyle) style, feature, (float) center.X, (float) center.Y);
            }
```

Changing the PolygonRender to use screen coordinates makes the label show up.
```
    internal static class PolygonRenderer
    {
        public static void Draw(SKCanvas canvas, IViewport viewport, IStyle style, IFeature feature, IGeometry geometry)
        {
            if (style is LabelStyle)
            {
                var center = geometry.GetBoundingBox().GetCentroid();
                var destination = viewport.WorldToScreen(center);
                LabelRenderer.Draw(canvas, (LabelStyle)style, feature, (float)destination.X, (float)destination.Y);
            }
```

Reproduce: Change the labels sample like so.
```
        public static ILayer CreateLayer()
        {
            var memoryProvider = new MemoryProvider();

            var featureWithDefaultStyle = new Feature {Geometry = new Point(0, 0)};
            featureWithDefaultStyle.Styles.Add(new LabelStyle { Text = "Default Label" });
            memoryProvider.Features.Add(featureWithDefaultStyle);


            var featureWithRightAlignedStyle = new Feature {Geometry = Geometry.GeomFromText("POLYGON((0 -2000000, 1000000 -2000000, 1000000 -4000000, 0 -4000000, 0 -2000000))") };
            featureWithRightAlignedStyle.Styles.Add(new LabelStyle
            {
                Text = "Polygon",
                BackColor = new Brush(Color.Gray),
                HorizontalAlignment = LabelStyle.HorizontalAlignmentEnum.Center
            });
            memoryProvider.Features.Add(featureWithRightAlignedStyle);
            
            return new MemoryLayer {Name = "Points with labels", DataSource = memoryProvider};
        }
```

**Comment by pauldendulk** [comment]
Thanks @Kortdag for your excellent analysis. The code to reproduce it allows me to pick this up quickly. I will try to push a nuget package later tonight. Although it is 'kortdag' I hope it is still in time :)

**Comment by pauldendulk** [comment]
This prerelease should fix this issue: https://www.nuget.org/packages/Mapsui/1.3.1-beta2

**Comment by Kortdag** [comment]
Hi Paul

Wow that was fast! Keep up the good work on the project its very usefull to us!!!
Haha "Kortdag" was an old account from a conference last year so everything is fine ☺.

The fix itself works as expected.

There is another issue that I also encountered by downloading the latest source from github and running the samples yesterday.
Thought it might be an emulator issue but the new package has similar issue on a real device.

It´s a problem in the iOS MapControl.
To reproduce
1) The latest source download from the main page on Github
2) Run the iOS label sample
3) make a pan or zoom operation
4) Does not seem to make a difference if running on other device emulators – or on a real device.
[cid:image001.png@01D38A08.4D322620]
5) Please find stacktrace attached

The locations collection in In MapControl gets and out of bounds exception since there is only one item in the collection – two expected.
private static double GetRotation(List<CGPoint> locations)
        {
            return Math.Atan2(locations[1].Y - locations[0].Y, locations[1].X - locations[0].X) * 180.0 / Math.PI;
        }


Caller is the method TouchesBegan.
public override void TouchesBegan(NSSet touches, UIEvent evt)
       {
            var locations = touches.Select(t => ((UITouch)t).LocationInView(this)).ToList();
            _previousRotation = GetRotation(locations);
            _touchDown = GetScreenPosition(touches);
            base.TouchesBegan(touches, evt);
        }


So I reverting back to the non beta verision for now. But thanks again for the swift reply.

Best regards
Simon

**Comment by spaddlewit** [comment]
Hey @Kortdag , do you get the pan bug ever since v1.2.0? It seems that the same thing happens to me on iOS, starting with v1.2.0 -- if I do a pan operation, the application crashes a split second afterward.

v1.1.3 works OK for me, so I'm sticking with that for now...

**Comment by pauldendulk** [comment]
I think it is something different. The issue kortdag mentioned was introduced the day before yesterday when I committed an update to rotation functionality in haste. I will fix that today.

**Comment by Kortdag** [comment]
I can confirm. We´re running 1.3.0 where pan works well.

The swift bugfixes is highly appreciated Paul!!

**Comment by pauldendulk** [comment]
@spaddlewit Could you submit a new issue for your problem?

**Comment by pauldendulk** [comment]
New beta release:
https://www.nuget.org/packages/Mapsui/1.3.1-beta3

**Comment by pauldendulk** [comment]
This was fixed with 1.3.1-beta3. There is a separate issue for the pan bug spaddlewit mentioned.
